### ISSUE TITLE
Wrong "Overloaded function implementation does not accept all possible arguments"

### ISSUE THREAD

**Issue by dvarrazzo** [python/mypy#11004]
**Bug Report**

The following example (reduced [from psycopg 3](https://github.com/psycopg/psycopg/blob/9b1af7cd2b68c3b1d229a722d8efbef4b360c7a5/psycopg/psycopg/connection.py#L532-L563)) seems to be a false positive:

**To Reproduce**
```python
from typing import Any, Callable, Generic, overload, Optional, TypeVar, Tuple, Union

Row = TypeVar("Row")

class Connection(Generic[Row]):
    @overload
    def connect(
        self, *, row_factory: Callable[[], Row], **kwargs: Union[None, int, str]
    ) -> "Connection[Row]":
        ...

    @overload
    def connect(
        self, **kwargs: Union[None, int, str],
    ) -> "Connection[Tuple[Any, ...]]":
        ...

    def connect(
        self, *, row_factory: Optional[Callable[[], Row]] = None, **kwargs: Any,
    ) -> "Connection[Any]":
        raise NotImplementedError
```

The file is flagged with the error:
```
bug_mypy.py:18: error: Overloaded function implementation does not accept all possible arguments of signature 2  [misc]
```

**Your Environment**

- Mypy version used: 0.910
- Python version used: 3.8
- Operating system and version: Linux

**Comment by erictraut** [comment]
You may have already figured this out, but there is a simple workaround. If you change the second overload to include an explicit `row_factory` parameter, it will type check without errors.

```python
    @overload
    def connect(
        self,
        *,
        row_factory: None = None,
        **kwargs: Union[None, int, str],
    ) -> "Connection[Tuple[Any, ...]]":
        ...
```

**Comment by dvarrazzo** [comment]
@erictraut  No, I didn't notice that, thank you. It is interesting, however I prefer to introduce a `type: ignore` rather than a fake parameter.

Another thing that make the error go is to use `**kwargs: Any` in the second signature.

**Comment by dlax** [comment]
I think that's the issue I tried to resolve in https://github.com/python/mypy/pull/10390.

**Comment by dvarrazzo** [comment]
Amazingg @dlax, didn't know you had looked into it so much to propose a MR! :slightly_smiling_face:

**Comment by adam-grant-hendry** [comment]
@dlax Do you plan to merge the change soon?

**Comment by dlax** [comment]
@adam-grant-hendry, the PR is still pending review; and then possible merge by a project member.

**Comment by sterliakov** [comment]
This isn't a bug? As currently written, the second overload accepts `Connection().connect(row_factory=10)` and the implementation signature does not, so `mypy` warns about a real problem.
